### PR TITLE
asan: slab arena and slab cache implementation

### PR DESCRIPTION
**Pull request by nshy**
Factored out from PR https://github.com/tarantool/small/pull/65 See it for a description.

Part of https://github.com/tarantool/tarantool/issues/7327

Tarantool PR for testing in CI is https://github.com/tarantool/tarantool/pull/8901. Note that testing PR tests all commits from https://github.com/tarantool/small/pull/65 but https://github.com/tarantool/small/pull/65 based on commits from this PR.

### CODE REVIEW

File: include/small/slab_cache_asan.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+static inline void
+slab_cache_check(struct slab_cache *cache)
+{
+	(void)cache;
+}
+
+static inline void
+slab_cache_set_thread(struct slab_cache *cache)
+{
+	(void)cache;
```
Comment (locker) [inline include/small/slab_cache_asan.h:116]: Why not implement the thread check for the asan version?

Comment (nshy) [reply include/small/slab_cache_asan.h:116]: Added check similar to non-ASAN version.

<details>
<summary>diff</summary>

```diff
diff --git a/include/small/slab_cache.h b/include/small/slab_cache.h
index b5e8aa6..078d86c 100644
--- a/include/small/slab_cache.h
+++ b/include/small/slab_cache.h
@@ -31,6 +31,7 @@
  * SUCH DAMAGE.
  */
 #include "slab_arena.h"
+#include <pthread.h>
 
 #include "small_config.h"
 
@@ -45,7 +46,6 @@
 #include <stddef.h>
 #include <assert.h>
 #include "rlist.h"
-#include <pthread.h>
 #include "slab_list.h"
 #include "util.h"
 
diff --git a/include/small/slab_cache_asan.h b/include/small/slab_cache_asan.h
index 7daaa2e..72c4acf 100644
--- a/include/small/slab_cache_asan.h
+++ b/include/small/slab_cache_asan.h
@@ -60,6 +60,9 @@ struct slab_cache {
 	size_t used;
 	/** Unique id among all allocators. */
 	unsigned int id;
+#ifndef NDEBUG
+	pthread_t thread_id;
+#endif
 };
 
 /** Header of slab cache allocation. */
@@ -114,6 +117,9 @@ static inline void
 slab_cache_set_thread(struct slab_cache *cache)
 {
 	(void)cache;
+#ifndef NDEBUG
+	cache->thread_id = pthread_self();
+#endif
 }
 
 #if defined(__cplusplus)
diff --git a/small/slab_cache_asan.c b/small/slab_cache_asan.c
index aa18190..31107d7 100644
--- a/small/slab_cache_asan.c
+++ b/small/slab_cache_asan.c
@@ -41,6 +41,7 @@ slab_cache_create(struct slab_cache *cache, struct slab_arena *arena)
 SMALL_NO_SANITIZE_ADDRESS struct slab *
 slab_get(struct slab_cache *cache, size_t size)
 {
+	assert(pthread_equal(cache->thread_id, pthread_self()));
 	size = slab_real_size(cache, size);
 	struct slab_cache_object *obj =
 			small_asan_alloc(size,
@@ -56,6 +57,7 @@ slab_get(struct slab_cache *cache, size_t size)
 SMALL_NO_SANITIZE_ADDRESS void
 slab_put(struct slab_cache *cache, struct slab *slab)
 {
+	assert(pthread_equal(cache->thread_id, pthread_self()));
 	struct slab_cache_object *obj = small_asan_header_from_payload(slab);
 	small_asan_assert(obj->cache_id == cache->id &&
 			  "object and cache id mismatch");

```

</details>

Comment (locker) [reply include/small/slab_cache_asan.h:119]: Let's enable it for release builds as well using `slab_asan_assert`.

Comment (nshy) [reply include/small/slab_cache_asan.h:119]: <details>
<summary>new diff</summary>

```diff
diff --git a/include/small/slab_cache.h b/include/small/slab_cache.h
index b5e8aa6..078d86c 100644
--- a/include/small/slab_cache.h
+++ b/include/small/slab_cache.h
@@ -31,6 +31,7 @@
  * SUCH DAMAGE.
  */
 #include "slab_arena.h"
+#include <pthread.h>
 
 #include "small_config.h"
 
@@ -45,7 +46,6 @@
 #include <stddef.h>
 #include <assert.h>
 #include "rlist.h"
-#include <pthread.h>
 #include "slab_list.h"
 #include "util.h"
 
diff --git a/include/small/slab_cache_asan.h b/include/small/slab_cache_asan.h
index 7daaa2e..913b4f5 100644
--- a/include/small/slab_cache_asan.h
+++ b/include/small/slab_cache_asan.h
@@ -60,6 +60,7 @@ struct slab_cache {
 	size_t used;
 	/** Unique id among all allocators. */
 	unsigned int id;
+	pthread_t thread_id;
 };
 
 /** Header of slab cache allocation. */
@@ -114,6 +115,7 @@ static inline void
 slab_cache_set_thread(struct slab_cache *cache)
 {
 	(void)cache;
+	cache->thread_id = pthread_self();
 }
 
 #if defined(__cplusplus)
diff --git a/small/slab_cache_asan.c b/small/slab_cache_asan.c
index aa18190..af89995 100644
--- a/small/slab_cache_asan.c
+++ b/small/slab_cache_asan.c
@@ -36,11 +36,13 @@ slab_cache_create(struct slab_cache *cache, struct slab_arena *arena)
 {
 	quota_lessor_create(&cache->quota, arena->quota);
 	cache->id = small_asan_reserve_id();
+	slab_cache_set_thread(cache);
 }
 
 SMALL_NO_SANITIZE_ADDRESS struct slab *
 slab_get(struct slab_cache *cache, size_t size)
 {
+	small_asan_assert(pthread_equal(cache->thread_id, pthread_self()));
 	size = slab_real_size(cache, size);
 	struct slab_cache_object *obj =
 			small_asan_alloc(size,
@@ -56,6 +58,7 @@ slab_get(struct slab_cache *cache, size_t size)
 SMALL_NO_SANITIZE_ADDRESS void
 slab_put(struct slab_cache *cache, struct slab *slab)
 {
+	small_asan_assert(pthread_equal(cache->thread_id, pthread_self()));
 	struct slab_cache_object *obj = small_asan_header_from_payload(slab);
 	small_asan_assert(obj->cache_id == cache->id &&
 			  "object and cache id mismatch");
```

</details>

**Review by locker — APPROVED** [review]

**Pull request merged**